link/trace_event.go

```diff
@@ -0,0 +1,322 @@
+//   exported kernel symbols. kprobe-based (tracefs) trace events can be
+//   created system-wide by writing to the <tracefs>/kprobe_events file, or
+//   they can be scoped to the current process by creating PMU perf events.
+// - perf event: An object instantiated based on an existing trace event or
+//   kernel symbol. Referred to by fd in userspace.
+//   Exactly one eBPF program can be attached to a perf event. Multiple perf
+//   events can be created from a single trace event. Closing a perf event
+//   stops any further invocations of the attached eBPF program.
+
+var (
+	tracefsPath = "/sys/kernel/debug/tracing"
+
+	// Trace event groups, names and kernel symbols must adhere to this set
+	// of characters. Non-empty, first character must not be a number, all
+	// characters must be alphanumeric or underscore.
+	rgxTraceEvent = regexp.MustCompile("^[a-zA-Z_][0-9a-zA-Z_]*$")
+)
+
+const (
+	perfAllThreads = -1
+)
+
+// A TraceEvent represents a trace event in the Linux kernel.
+type TraceEvent struct {
+	// Group and name of the tracepoint/kprobe/uprobe.
+	group string
+	name  string
+
+	// PMU event ID read from sysfs. Valid IDs are non-zero.
+	pmuID uint64
+	// ID of the trace event read from tracefs. Valid IDs are non-zero.
+	tracefsID uint64
+
+	// True for kretprobes/uretprobes.
+	ret bool
+
+	// Type of program this trace event will accept to be attached.
+	progType ebpf.ProgramType
+}
+
+// Close releases the resources associated with the TraceEvent.
+func (te *TraceEvent) Close() error {
+	if te.group == "" || te.name == "" {
+		return errors.New("cannot close an empty trace event")
+	}
+
+	switch t := te.progType; t {
+	case ebpf.Kprobe:
+		return te.closeKprobe()
+	case ebpf.TracePoint:
+		return nil
+	default:
+		return fmt.Errorf("unknown program type for trace event: %s", t)
+	}
+}
+
+// Attach opens a new perf event for the TraceEvent and attaches the given
+// program to it.
+func (te *TraceEvent) Attach(program *ebpf.Program) (Link, error) {
+	if program == nil {
+		return nil, errors.New("cannot attach a nil program")
+	}
+	if t := program.Type(); t != te.progType {
+		return nil, fmt.Errorf("invalid program type (expected %s): %s", te.progType, t)
+	}
+	if program.FD() < 0 {
+		return nil, fmt.Errorf("invalid program: %w", internal.ErrClosedFd)
+	}
+
+	var ifd *internal.FD
+
+	switch {
+	case te.pmuID != 0:
+		// Open a PMU perf event based on a sysfs PMU type.
+		fd, err := te.perfEventOpenPMU()
+		if err != nil {
+			return nil, fmt.Errorf("open PMU trace event: %w", err)
+		}
+		ifd = fd
+
+	case te.tracefsID != 0:
+		// Open a perf event based on a tracefs trace event ID.
+		fd, err := te.perfEventOpenTraceFS()
+		if err != nil {
+			return nil, fmt.Errorf("open tracefs perf event: %w", err)
+		}
+		ifd = fd
+
+	default:
+		return nil, errors.New("need a PMU type or tracefs ID")
+	}
+
+	kfd, err := ifd.Value()
+	if err != nil {
+		return nil, fmt.Errorf("getting perf event fd: %w", err)
```

Comment:
This and the following return leak `ifd`.

Nit: I think it'd be ok to ignore the error here and just use kfd.

Reply:
Good catch. fixed.